db/src/metadata.rs

```diff
@@ -0,0 +1,379 @@
+                // but during bootstrapping, we are given complete attributes (including
+                // :db/valueType) that are already in the schema.  This allows to ignore that case.
+                let mutations = builder.mutate(entry.get_mut());
+                if mutations.is_empty() {
+                    continue
+                }
+                if !builder.is_valid_alter_attribute() {
+                    bail!(ErrorKind::BadSchemaAssertion(format!("Schema attribute for :db.alter/attribute with entid '{}' must not set :db/valueType", entid)));
+                }
+                mutated.insert(entid, mutations);
+            },
+        }
+    }
+
+    Ok(MetadataReport {
+        flagged_install: installed,
+        flagged_alter: altered,
+        installed: asserted,
+        altered: mutated,
+    })
+}
+
+/// Update a `Schema` in place from the given `[e a typed_value added]` quadruples.
+///
+/// This layer enforces that [:db.part/db :db.install/attribute ...] and [:db.part/db
+/// :db.alter/attribute ...] markers are present and correct.
+///
+/// This is suitable for mutating a `Schema` from an applied transaction.
+///
+/// Returns a report summarizing the mutations that were applied.
+pub fn update_metadata_from_committed_metadata_assertions<U>(schema: &mut Schema, assertions: U) -> Result<MetadataReport>
+    where U: IntoIterator<Item=(Entid, Entid, TypedValue, bool)> {
+
+    // Group into asserted, retracted, and updated.  We assume all our attributes are
+    // :db/cardinality :db.cardinality/one (so they'll only be added or retracted at most once),
+    // and all modifications are simple changes.
+    let mut asserted: BTreeMap<(Entid, Entid), TypedValue> = BTreeMap::new();
+    let mut retracted: BTreeMap<(Entid, Entid), TypedValue> = BTreeMap::new();
+    let mut updated: BTreeMap<(Entid, Entid), (TypedValue, TypedValue)> = BTreeMap::new();
+
+    for (e, a, typed_value, added) in assertions.into_iter() {
+        let attribute = schema.require_attribute_for_entid(a)?;
+        // XXX TODO NOTE
+        if attribute.multival {
+            continue
+            // bail!(ErrorKind::NotYetImplemented(format!("Handling metadata assertions with attribute {} with :db/cardinality :db.cardinality/many not yet implemented", a)));
+        }
+
+        if added {
+            if let Some(retracted_value) = retracted.remove(&(e, a)) {
+                updated.insert((e, a), (retracted_value, typed_value));
+            } else {
+                asserted.insert((e, a), typed_value);
+            }
+        } else {
+            if let Some(asserted_value) = asserted.remove(&(e, a)) {
+                updated.insert((e, a), (typed_value, asserted_value));
+            } else {
+                retracted.insert((e, a), typed_value);
+            }
+        }
+    }
+
+    // Datomic does not allow to retract attributes or idents.  For now, Mentat follows suit.
```

Comment:
The schema doc suggests you can retract _something_ to unset things like uniqueness constraints.